File: 2024/python/day9/solution.py
from typing import List, Literal


def move_train(
    board: List[str], mov: Literal["U", "D", "R", "L"]
) -> Literal["none", "crash", "eat"]:
    movements = {
        "U": (-1, 0),
        "D": (1, 0),
        "L": (0, -1),
        "R": (0, 1),
    }

    head_x, head_y = None, None
    for i, row in enumerate(board):
        if "@" in row:
            head_x, head_y = i, row.index("@")
            break

    dx, dy = movements[mov]
    new_x, new_y = head_x + dx, head_y + dy

    if new_x < 0 or new_x >= len(board) or new_y < 0 or new_y >= len(board[0]):
        return "crash"

    next_cell = board[new_x][new_y]
    if next_cell == "o":
        return "crash"
    if next_cell == "*":
        return "eat"

    return "none"

File: 2024/js/day9/solution.js
/**
 * @param {string[]} board - List of strings representing the board.
 * @param {"U" | "D" | "R" | "L"} mov - The movement direction.
 * @returns {"none" | "crash" | "eat"} - The resulting state of the train.
 */
function moveTrain(board, mov) {
  const movements = {
    U: [-1, 0],
    D: [1, 0],
    L: [0, -1],
    R: [0, 1],
  };

  let headX = null,
    headY = null;

  for (let i = 0; i < board.length; i++) {
    const row = board[i];
    const index = row.indexOf("@");
    if (index !== -1) {
      headX = i;
      headY = index;
      break;
    }
  }

  const [dx, dy] = movements[mov];
  const newX = headX + dx;
  const newY = headY + dy;

  if (newX < 0 || newX >= board.length || newY < 0 || newY >= board[0].length) {
    return "crash";
  }

  const nextCell = board[newX][newY];

  if (nextCell === "o") {
    return "crash";
  }
  if (nextCell === "*") {
    return "eat";
  }

  return "none";
}
